### ISSUE TITLE
Deprecate unused module attributes in twisted.protocols.basic.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4541]
|<img alt="brandl's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| brandl reported|
|-|-|
|Trac ID|trac#4541|
|Type|defect|
|Created|2010-07-02 10:38:47Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-old-netstring-attributes-4541|

Due to changes in `t.p.b.NetstringReceiver`, the module attributes
`LENGTH`, `DATA`, `COMMA`, and `NUMBER` are no longer used.

Please deprecate them.

Attachments:

* [ticket_4541.1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c1e/c1ece8a30d07e86687eeb87ebf56b82d6a26f8af/584e81f8afbebdeb71b8a470ff0ae7302938fa1a.patch) (1292 bytes) - added by brandl on 2010-07-14 09:03:26Z - A patch against the trunk deprecating the attributes LENGTH, DATA, COMMA, and NUMBER
* [ticket_4541.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c1e/c1ece8a30d07e86687eeb87ebf56b82d6a26f8af/6eb922bb8fa2a852c3617d90d342a6fb22be936a.patch) (3847 bytes) - added by brandl on 2010-07-21 10:46:06Z - Added news entry (forgot to call "svn add" :-( )

<details><summary>Searchable metadata</summary>

```
trac-id__4541 4541
type__defect defect
reporter__brandl brandl
priority__normal normal
milestone__ 
branch__branches_deprecate_old_netstring_attributes_4541 branches/deprecate-old-netstring-attributes-4541
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1278067127000000 1278067127000000
changetime__1283562504000000 1283562504000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="brandl's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|brandl set owner to @exarkun|
|-|-|

I've attached a patch for deprecating these attributes. Not sure if the
deprecation version is correct, though.

Please check.

BTW: I'm not too familiar with the workflow of tickets - is it OK to
assign these things directly to you, or should I just change the keyword
to "review" and wait until someone grabs the issue?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [brandl](#note_1):
> BTW: I'm not too familiar with the workflow of tickets - is it OK to
> assign these things directly to you, or should I just change the keyword
> to "review" and wait until someone grabs the issue?

Thanks for the patch, see TwistedDevelopment for more info.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj set owner to brandl|
|-|-|

## Code Review
Looks mostly fine to me - tests pass, warnings are printed upon accessing those module vars.  

A few problems:
 0. According to the policy, you need to add unit tests for deprecations (see http://twistedmatrix.com/trac/wiki/CompatibilityPolicy#Unittesting)
 0. Missing news file (see http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles)
 0. I think deprecatedSince should be 10.1.1 (See http://twistedmatrix.com/trac/wiki/CompatibilityPolicy#Functionsandmethods "the version of Twisted in which it was first deprecated"
 0. I'd say you should either delete those left over module level private vars (and basic.attrib, basic.reason) afterwards, OR
 0. ....just keep it simple (and clearer) by writing three separate deprecate.deprecatedModuleAttribute calls instead of the loop.
 0. "NUMBER", _useInstead % ("_LENGTH",): I don't know NetstringReceiver very well, but is that the correct alternative?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> I think deprecatedSince should be 10.1.1

10.2 :)

**Comment by twisted-trac** [comment]
|<img alt="brandl's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|brandl set owner to @wallrj|
|-|-|

Thanks for the review. I've uploaded a second patch.

There is a new unit test checking for the deprecated module attributes and
I have added a news file. The code was simplified according to your suggestions,
and the `_deprecatedSince` attribute is deleted after its use now.

I've changed the version number for deprecatedSince to 10.2.0, following 
the comment by exarkun. I'm still not sure how to determine this number, though.
Is this the version of Twisted that includes the fix for 
http://twistedmatrix.com/trac/ticket/4378? Starting with this version, the
module attributes should not be used anymore.

Is there a simple way to determine in which version ticket 4378 will be 
fixed? I didn't find it in the tickets for Milestone 10.2, and there is
no Milestone for 10.1.1.

Oh, and yes: `NUMBER` was replaced by `NetstringReceiver._LENGTH`. They both
represent a match object for parsing the length specification of a netstring.
Maybe `_LENGTH_PATTERN` would have been a better name?

Please check.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

Replying to [brandl](#note_5):
> Thanks for the review. I've uploaded a second patch.
> There is a new unit test checking for the deprecated module attributes and
> I have added a news file. The code was simplified according to your suggestions,
> and the `_deprecatedSince` attribute is deleted after its use now.

Great, it looks much clearer to me.

> I've changed the version number for deprecatedSince to 10.2.0, following 
> the comment by exarkun. I'm still not sure how to determine this number, though.
> Is this the version of Twisted that includes the fix for 
> http://twistedmatrix.com/trac/ticket/4378? Starting with this version, the
> module attributes should not be used anymore.
> Is there a simple way to determine in which version ticket 4378 will be 
> fixed? I didn't find it in the tickets for Milestone 10.2, and there is
> no Milestone for 10.1.1.

Not sure. I'll leave someone else to answer that.

> Oh, and yes: `NUMBER` was replaced by `NetstringReceiver._LENGTH`. They both
> represent a match object for parsing the length specification of a netstring.
> Maybe `_LENGTH_PATTERN` would have been a better name?

Understood.

> Please check.

Looks good. Tests pass. Only one thing:

```
richard@aziz:~/Projects/Twisted/trunk$ pyflakes twisted/test/test_protocols.py
twisted/test/test_protocols.py:614: local variable 'length' is assigned to but never used
twisted/test/test_protocols.py:625: local variable 'length' is assigned to but never used
```

...which may be annoying for anyone looking at the automated pyflakes buildbot output. I guess you don't need to assign the module variables in order to trigger the warning.

You'll have to get one of the twisted devs (eg exarkun) to merge this code to trunk, so see what they say about this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun|
|-|-|

  * For deprecations, please use a .removal news fragment.
  * The deprecation warnings should not point people to private attributes.  There is no public replacement for these attributes, that's fine.
  * Indeed, the deprecation warning can be tested just by reading the attribute, no local needs to be assigned to.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29957]) Branching to 'deprecate-old-netstring-attributes-4541'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29958]) apply ticket_4541.2.patch

refs [#4541](https://github.com/twisted/twisted/issues/4541)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [29959]) Change the deprecation messages to not point people at implementation details and just warn them away instead.

refs [#4541](https://github.com/twisted/twisted/issues/4541)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [29960]) Fix two pyflakes warnings by adding some asserts about the return value of _consumeLength

refs [#4541](https://github.com/twisted/twisted/issues/4541)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [29961]) Oops; no, _consumeLength returns None

refs [#4541](https://github.com/twisted/twisted/issues/4541)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [29962]) Merge deprecate-old-netstring-attributes-4541

Author: brandl, exarkun
Reviewer: rwall, exarkun
Fixes: [#4541](https://github.com/twisted/twisted/issues/4541)

Deprecate the `twisted.protocols.basic` module attributes which were only
in support of the old `NetstringReceiver` parser (no longer in use since
[#4378](https://github.com/twisted/twisted/issues/4378) was resolved).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
